mecanum_drive_controller/test/test_mecanum_drive_controller.cpp

```diff
@@ -0,0 +1,547 @@
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  // EXPECT_EQ(joint_command_values_[NR_CMD_ITFS - 2], TEST_LINEAR_VELOCITY_X);
+  EXPECT_FALSE(std::isnan(joint_command_values_[1]));
+  EXPECT_NE(joint_command_values_[1], 111);
+  EXPECT_EQ(joint_command_values_[1], 3.0);
+
+  ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->twist.linear.x, TEST_LINEAR_VELOCITY_X);
+}
+
+TEST_F(MecanumDriveControllerTest, test_ref_timeout_zero_for_update)
+{
+  SetUpController();
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(controller_->get_node()->get_node_base_interface());
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  auto reference = controller_->input_ref_.readFromNonRT();
+
+  // set command statically
+  static constexpr double TEST_LINEAR_VELOCITY_X = 1.5;
+  static constexpr double TEST_LINEAR_VELOCITY_y = 0.0;
+  static constexpr double TEST_ANGULAR_VELOCITY_Z = 0.0;
+  joint_command_values_[1] = 111;
+
+  controller_->ref_timeout_ = rclcpp::Duration::from_seconds(0.0);
+  std::shared_ptr<ControllerReferenceMsg> msg = std::make_shared<ControllerReferenceMsg>();
+
+  msg->header.stamp = controller_->get_node()->now() - rclcpp::Duration::from_seconds(0.0);
+  msg->twist.linear.x = TEST_LINEAR_VELOCITY_X;
+  msg->twist.linear.y = TEST_LINEAR_VELOCITY_y;
+  msg->twist.linear.z = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.x = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.y = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.z = TEST_ANGULAR_VELOCITY_Z;
+  controller_->input_ref_.writeFromNonRT(msg);
+  const auto age_of_last_command =
+    controller_->get_node()->now() - (*(controller_->input_ref_.readFromNonRT()))->header.stamp;
+
+  ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->twist.linear.x, TEST_LINEAR_VELOCITY_X);
+  ASSERT_EQ(
+    controller_->update_reference_from_subscribers(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  // EXPECT_EQ(joint_command_values_[NR_STATE_ITFS - 2], TEST_LINEAR_VELOCITY_X);
+  EXPECT_FALSE(std::isnan(joint_command_values_[1]));
+  EXPECT_NE(joint_command_values_[1], 111);
+  EXPECT_EQ(joint_command_values_[1], 3.0);
+  ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->twist.linear.x, TEST_LINEAR_VELOCITY_X);
+  ASSERT_FALSE(std::isnan((*(controller_->input_ref_.readFromRT()))->twist.linear.x));
+}
+
+TEST_F(MecanumDriveControllerTest, test_ref_timeout_zero_for_reference_callback)
```

Comment:
what is the purpose of this test?

Reply:
to test the behavior of reference_callback() when ref_timeout = 0

Reply:
And what is here different compared to the above behavior?

Reply:
the above test is testing ref_timeout for update()

calling update implicitly calls, ref_callback, writes to ref_interfces, computes kinematics and writes to command_interfaces

calling ref_callback alone does only one job, writing fron nonRT thread to rtbuffer.